internal/command/deploy/machinebasedtest.go

```diff
@@ -0,0 +1,256 @@
+				md.colorize.Bold(testMachine.Machine().ID), md.colorize.Red(strconv.Itoa(exitCode)))
+			fmt.Fprintf(md.io.ErrOut, "Check its logs: here's the last 100 lines below, or run 'fly logs -i %s':\n",
+				testMachine.Machine().ID)
+			testLogs, _, err := md.apiClient.GetAppLogs(ctx, md.app.Name, "", md.appConfig.PrimaryRegion, testMachine.Machine().ID)
+			if fly.IsNotAuthenticatedError(err) {
+				fmt.Fprintf(md.io.ErrOut, "Warn: not authorized to retrieve app logs (this can happen when using deploy tokens), so we can't show you what failed. Use `fly logs -i %s` or open the monitoring dashboard to see them: https://fly.io/apps/%s/monitoring?region=&instance=%s\n", testMachine.Machine().ID, md.appConfig.AppName, testMachine.Machine().ID)
+			} else {
+				if err != nil {
+					return fmt.Errorf("Error getting test command logs: %w", err)
+				}
+				for _, l := range testLogs {
+					fmt.Fprintf(md.io.ErrOut, "  %s\n", l.Message)
+				}
+			}
+			return fmt.Errorf("Error test command machine %s exited with non-zero status of %d", testMachine.Machine().ID, exitCode)
+		}
+		statuslogger.LogfStatus(ctx,
+			statuslogger.StatusSuccess,
+			"Test machine %s completed successfully",
+			md.colorize.Bold(testMachine.Machine().ID),
+		)
+	}
+
+	return nil
+}
+
+const ErrNoLogsFound = "no logs found"
+
+func (md *machineDeployment) waitForLogs(ctx context.Context, mach *fly.Machine, timeout time.Duration) error {
+	b := backoff.NewExponentialBackOff()
+	b.InitialInterval = 1 * time.Second
+	b.MaxInterval = 10 * time.Second
+	b.MaxElapsedTime = timeout
+
+	return backoff.Retry(func() error {
+		logs, _, err := md.apiClient.GetAppLogs(ctx, md.app.Name, "", md.appConfig.PrimaryRegion, mach.ID)
+		if err != nil {
+			return err
+		}
+		if len(logs) == 0 {
+			return fmt.Errorf(ErrNoLogsFound)
+		}
+
+		return nil
+	}, backoff.WithContext(b, ctx))
+}
+
+func (md *machineDeployment) createTestMachine(ctx context.Context, svc *appconfig.ServiceMachineCheck, machineToTest *fly.Machine) (*fly.Machine, error) {
+	ctx, span := tracing.GetTracer().Start(ctx, "create_test_machine")
+	defer span.End()
+
+	launchInput, err := md.launchInputForTestMachine(svc, machineToTest)
+	if err != nil {
+		return nil, err
+	}
+	testMachine, err := md.flapsClient.Launch(ctx, *launchInput)
+	if err != nil {
+		tracing.RecordError(span, err, "failed to create test machines")
+		return nil, fmt.Errorf("error creating a test machine: %w", err)
+	}
+
+	statuslogger.Logf(ctx, "Created test machine %s", md.colorize.Bold(testMachine.ID))
+	return testMachine, nil
+}
+
+func (md *machineDeployment) launchInputForTestMachine(svc *appconfig.ServiceMachineCheck, origMachineRaw *fly.Machine) (*fly.LaunchMachineInput, error) {
+	if origMachineRaw == nil {
+		origMachineRaw = &fly.Machine{
+			Region: md.appConfig.PrimaryRegion,
+		}
+	}
+
+	mConfig, err := md.appConfig.ToTestMachineConfig(svc, origMachineRaw)
+	if err != nil {
+		return nil, err
+	}
+	mConfig.Guest = md.inferTestMachineGuest()
```

Comment:
having another guest inferring function here sucks (sorry), what about using the same machine size than the canary? or even add a `appconfig.Compute` sub section to machine checks.

Reply:
yea i don't mind just using `inferCanaryGuest` here